jni/themis_message.c

```diff
@@ -155,35 +153,50 @@ JNIEXPORT jbyteArray JNICALL Java_com_cossacklabs_themis_SecureMessage_process(J
                                             public_key_length,
                                             (uint8_t*)message_buf,
                                             message_length,
-                                            (uint8_t*)output_buf,
+                                            output_buffer,
                                             &output_length);
         break;
     case SECURE_MESSAGE_SIGN:
         res = themis_secure_message_sign((uint8_t*)priv_buf,
                                          private_key_length,
                                          (uint8_t*)message_buf,
                                          message_length,
-                                         (uint8_t*)output_buf,
+                                         output_buffer,
                                          &output_length);
         break;
     case SECURE_MESSAGE_VERIFY:
         res = themis_secure_message_verify((uint8_t*)pub_buf,
                                            public_key_length,
                                            (uint8_t*)message_buf,
                                            message_length,
-                                           (uint8_t*)output_buf,
+                                           output_buffer,
                                            &output_length);
         break;
     default:
         res = THEMIS_NOT_SUPPORTED;
         break;
     }
 
-err:
+    if (res != THEMIS_SUCCESS) {
+        goto err;
+    }
 
-    if (output_buf) {
-        (*env)->ReleaseByteArrayElements(env, output, output_buf, 0);
+    output = (*env)->NewByteArray(env, output_length);
+    if (!output) {
+        res = THEMIS_NO_MEMORY;
+        goto err;
+    }
+
+    output_bytes = (*env)->GetPrimitiveArrayCritical(env, output, NULL);
```

Comment:
so we pause garbage collectors here, moves bytes from `output_buffer` to `output_bytes`, and un-pause garbage collectors, right?

i followed description of `GetPrimitiveArrayCritical ` at [SO](https://stackoverflow.com/a/46608121).

Reply:
> so we pause garbage collectors here, moves bytes from output_buffer to output_bytes, and un-pause garbage collectors, right?

Basically yeah. Either pause to synchronize access, or making a copy and synchronizing copying it back.

This may or may not have an effect on the overall system. I'm sure there are effects to using both critical and non-critical variants, but I'm not an expert on Dalvik or other JVMs so have no clue on how strong those effects can be.